// pinchtab/dist/src/download.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensureBinary = ensureBinary;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const https = __importStar(require("https"));
const crypto = __importStar(require("crypto"));
const platform_1 = require("./platform");
const GITHUB_REPO = 'pinchtab/pinchtab';
// Read version from package.json at build time
function getVersion() {
    const pkgPath = path.join(__dirname, '..', 'package.json');
    const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'));
    return pkg.version;
}
function fetchUrl(url, maxRedirects = 5) {
    return new Promise((resolve, reject) => {
        const attemptFetch = (currentUrl, redirectsRemaining) => {
            let agent;
            // Proxy support for corporate environments
            const proxyUrl = process.env.HTTPS_PROXY || process.env.HTTP_PROXY;
            if (proxyUrl) {
                try {
                    const proxy = new URL(proxyUrl);
                    const proxyPort = proxy.port ? parseInt(proxy.port, 10) : 8080;
                    agent = new https.Agent({
                        host: proxy.hostname,
                        port: proxyPort,
                        keepAlive: true,
                    });
                }
                catch (_err) {
                    console.warn(`Warning: Invalid proxy URL ${proxyUrl}, ignoring`);
                }
            }
            const request = https.get(currentUrl, agent ? { agent } : {}, (response) => {
                // Handle redirects (301, 302, 307, 308)
                if ([301, 302, 307, 308].includes(response.statusCode || 0)) {
                    if (redirectsRemaining <= 0) {
                        reject(new Error(`Too many redirects from ${currentUrl}`));
                        return;
                    }
                    let redirectUrl = response.headers.location;
                    if (!redirectUrl) {
                        reject(new Error(`Redirect without location header from ${currentUrl}`));
                        return;
                    }
                    // Resolve relative URLs
                    try {
                        redirectUrl = new URL(redirectUrl, currentUrl).toString();
                    }
                    catch (_err) {
                        reject(new Error(`Invalid redirect URL from ${currentUrl}: ${redirectUrl}`));
                        return;
                    }
                    // Consume the response stream to avoid memory leaks
                    response.resume();
                    attemptFetch(redirectUrl, redirectsRemaining - 1);
                    return;
                }
                if (response.statusCode === 404) {
                    reject(new Error(`Not found: ${currentUrl}`));
                    return;
                }
                if (response.statusCode !== 200) {
                    reject(new Error(`HTTP ${response.statusCode}: ${currentUrl}`));
                    return;
                }
                const chunks = [];
                response.on('data', (chunk) => chunks.push(chunk));
                response.on('end', () => resolve(Buffer.concat(chunks)));
                response.on('error', reject);
            });
            request.on('error', reject);
        };
        attemptFetch(url, maxRedirects);
    });
}
async function downloadChecksums(version) {
    const url = `https://github.com/${GITHUB_REPO}/releases/download/v${version}/checksums.txt`;
    try {
        const data = await fetchUrl(url);
        const checksums = new Map();
        data
            .toString('utf-8')
            .trim()
            .split('\n')
            .forEach((line) => {
            const [hash, filename] = line.split(/\s+/);
            if (hash && filename) {
                checksums.set(filename.trim(), hash.trim());
            }
        });
        return checksums;
    }
    catch (err) {
        throw new Error(`Failed to download checksums: ${err.message}. ` +
            `Ensure v${version} is released on GitHub with checksums.txt`);
    }
}
function verifySHA256(filePath, expectedHash) {
    const hash = crypto.createHash('sha256');
    const data = fs.readFileSync(filePath);
    hash.update(data);
    const actualHash = hash.digest('hex');
    return actualHash.toLowerCase() === expectedHash.toLowerCase();
}
async function downloadBinary(platform, version) {
    const binaryName = (0, platform_1.getBinaryName)(platform);
    const binaryPath = (0, platform_1.getBinaryPath)(binaryName, version);
    // Always verify existing binaries, even if they exist
    // (guards against corrupted installs from previous failures)
    if (fs.existsSync(binaryPath)) {
        try {
            const checksums = await downloadChecksums(version);
            if (checksums.has(binaryName)) {
                const expectedHash = checksums.get(binaryName);
                if (verifySHA256(binaryPath, expectedHash)) {
                    console.log(`✓ Pinchtab binary verified: ${binaryPath}`);
                    return;
                }
                else {
                    console.warn(`⚠ Existing binary failed checksum, re-downloading...`);
                    fs.unlinkSync(binaryPath);
                }
            }
        }
        catch (_err) {
            console.warn(`⚠ Could not verify existing binary, re-downloading...`);
            try {
                fs.unlinkSync(binaryPath);
            }
            catch {
                // ignore
            }
        }
    }
    // Fetch checksums
    console.log(`Downloading Pinchtab ${version} for ${platform.os}-${platform.arch}...`);
    const checksums = await downloadChecksums(version);
    if (!checksums.has(binaryName)) {
        throw new Error(`Binary not found in checksums: ${binaryName}. ` +
            `Available: ${Array.from(checksums.keys()).join(', ')}`);
    }
    const expectedHash = checksums.get(binaryName);
    const downloadUrl = `https://github.com/${GITHUB_REPO}/releases/download/v${version}/${binaryName}`;
    // Ensure directory exists (unless using custom PINCHTAB_BINARY_PATH)
    const binDir = path.dirname(binaryPath);
    if (!fs.existsSync(binDir)) {
        fs.mkdirSync(binDir, { recursive: true });
    }
    // Download to temp file first, then atomically rename to final path
    // This prevents partial/corrupted files from being left behind
    const tempPath = `${binaryPath}.tmp`;
    return new Promise((resolve, reject) => {
        console.log(`Downloading from ${downloadUrl}...`);
        const file = fs.createWriteStream(tempPath);
        let redirectCount = 0;
        const maxRedirects = 5;
        const performDownload = (url) => {
            https
                .get(url, (response) => {
                // Handle redirects (301, 302, 307, 308)
                if ([301, 302, 307, 308].includes(response.statusCode || 0)) {
                    if (redirectCount >= maxRedirects) {
                        fs.unlink(tempPath, () => { });
                        reject(new Error(`Too many redirects downloading ${downloadUrl}`));
                        return;
                    }
                    let redirectUrl = response.headers.location;
                    if (!redirectUrl) {
                        fs.unlink(tempPath, () => { });
                        reject(new Error(`Redirect without location header from ${url}`));
                        return;
                    }
                    // Resolve relative URLs
                    try {
                        redirectUrl = new URL(redirectUrl, url).toString();
                    }
                    catch (_err) {
                        fs.unlink(tempPath, () => { });
                        reject(new Error(`Invalid redirect URL from ${url}: ${redirectUrl}`));
                        return;
                    }
                    redirectCount++;
                    response.resume(); // Consume response stream
                    performDownload(redirectUrl);
                    return;
                }
                if (response.statusCode !== 200) {
                    fs.unlink(tempPath, () => { });
                    reject(new Error(`HTTP ${response.statusCode}: ${url}`));
                    return;
                }
                response.pipe(file);
                file.on('finish', () => {
                    file.close();
                    // Verify checksum before moving to final location
                    if (!verifySHA256(tempPath, expectedHash)) {
                        fs.unlink(tempPath, () => { });
                        reject(new Error(`Checksum verification failed for ${binaryName}. ` +
                            `Download may be corrupted. Please try again.`));
                        return;
                    }
                    // Atomically move temp file to final location
                    try {
                        fs.renameSync(tempPath, binaryPath);
                    }
                    catch (err) {
                        fs.unlink(tempPath, () => { });
                        reject(new Error(`Failed to finalize binary: ${err.message}`));
                        return;
                    }
                    // Make executable
                    try {
                        fs.chmodSync(binaryPath, 0o755);
                    }
                    catch (err) {
                        // On Windows, chmod may fail but binary may still be usable
                        console.warn(`⚠ Warning: could not set executable permissions: ${err.message}`);
                    }
                    console.log(`✓ Verified and installed: ${binaryPath}`);
                    resolve();
                });
                file.on('error', (err) => {
                    fs.unlink(tempPath, () => { });
                    reject(err);
                });
            })
                .on('error', reject);
        };
        performDownload(downloadUrl);
    });
}
async function ensureBinary() {
    const platform = (0, platform_1.detectPlatform)();
    const version = getVersion();
    await downloadBinary(platform, version);
    const binaryName = (0, platform_1.getBinaryName)(platform);
    return (0, platform_1.getBinaryPath)(binaryName, version);
}
