// pinchtab/dist/src/platform.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectPlatform = detectPlatform;
exports.getBinaryName = getBinaryName;
exports.getBinDir = getBinDir;
exports.getBinaryPath = getBinaryPath;
const path = __importStar(require("path"));
function detectPlatform() {
    const platform = process.platform;
    // Only support x64 (amd64) and arm64
    let arch;
    if (process.arch === 'x64') {
        arch = 'amd64';
    }
    else if (process.arch === 'arm64') {
        arch = 'arm64';
    }
    else {
        throw new Error(`Unsupported architecture: ${process.arch}. ` + `Only x64 (amd64) and arm64 are supported.`);
    }
    const osMap = {
        darwin: 'darwin',
        linux: 'linux',
        win32: 'windows',
    };
    const os_name = osMap[platform];
    if (!os_name) {
        throw new Error(`Unsupported platform: ${platform}`);
    }
    return { os: os_name, arch };
}
function getBinaryName(platform) {
    const { os, arch } = platform;
    const archName = arch === 'arm64' ? 'arm64' : 'amd64';
    if (os === 'windows') {
        return `pinchtab-${os}-${archName}.exe`;
    }
    return `pinchtab-${os}-${archName}`;
}
function getBinDir() {
    return path.join(process.env.HOME || process.env.USERPROFILE || '', '.pinchtab', 'bin');
}
function getBinaryPath(binaryName, version) {
    // Allow override via environment variable (for Docker, custom builds, etc.)
    if (process.env.PINCHTAB_BINARY_PATH) {
        return process.env.PINCHTAB_BINARY_PATH;
    }
    // Version-specific path: ~/.pinchtab/bin/0.7.0/pinchtab-darwin-arm64
    // This allows multiple versions to coexist and prevents silent overwrites
    if (version) {
        return path.join(getBinDir(), version, binaryName);
    }
    // Fallback to version-less for backwards compat
    return path.join(getBinDir(), binaryName);
}
